Keep BO from using two different translators in parallel

matthieu-rolland opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Fix translation issues, read below
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #14932
| How to test?  | Check that translation is still working, especially translations coming from database in modules.

First, you should read this PR from @eternoendless 
https://github.com/PrestaShop/PrestaShop/pull/14845

His PR renames translation test modules, allows underscores in translation domain name, and adds useful comments to understand the two translators.

From there, this PR adds a few things, the main goal being that wether we are in FO or BO, we always use **only one translator**, either the translator component or the one from frameworkbundle. Also make it so that we still have the translation part in symfony's debug toolbar, and that the database translations are still working in modules.

- make frameworkbundle's translator load SQL translations with yaml service configuration
- fix timing issue in controllerCore (service container was called after the first call to the translator) . (edit: removed due to issue with php 5.6)  
- Fetch the right translator depending on the availability or not of the symfony service container
- Make it possible to specify which translator you need

I carefully checked that the component translator is never called in back office, and that the framework bundle translator is never called in front office.


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/14966)
<!-- Reviewable:end -->

matthieu-rolland committed:
add sql translation loader in translation.yml

matthieu-rolland committed:
In controllerCore, set service container before the translator is called for the first time

matthieu-rolland committed:
Fetch the right translator depending on the availability of symfony container

mickaelandrieu commented on classes/controller/Controller.php:
```diff
@@ -180,9 +180,6 @@ public function init()
-        if (null === $this->getContainer()) {
-            $this->container = $this->buildContainer();
-        }
         $localeRepo = $this->get(self::SERVICE_LOCALE_REPOSITORY);
```
you call the container before its build, isn't it ?

matthieu-rolland replied:
yes, I call the container builder so that its built. That's what was done, I just moved this process from the init() method to the constructor.

mickaelandrieu commented on src/PrestaShopBundle/Translation/Translator.php:
```diff
@@ -29,6 +29,9 @@
 use Symfony\Bundle\FrameworkBundle\Translation\Translator as BaseTranslator;
 
+/**
+ * Replaces the original Symfony FrameworkBundle translator
```
I'm not sure this PHPDoc is true, it's the compiler pass that do the replacement.

eternoendless replied:
I didn't think it could be interpreted that way. Here's a better wording:

```suggestion
 * Replacement for the original Symfony FrameworkBundle translator
```

mickaelandrieu commented on tests/Integration/PrestaShopBundle/Translation/Provider/ExternalModuleLegacySystemProviderTest.php:
```diff
@@ -37,7 +37,7 @@
 class ExternalModuleLegacySystemProviderTest extends KernelTestCase
 {
-    const MODULE_NAME = 'translation_test';
+    const MODULE_NAME = 'translationtest';
```
underscore are valid in module name: why do we need to change that ?

matthieu-rolland replied:
As stated in the PR's description, you should really read @eternoendless 's PR . ;)

https://github.com/PrestaShop/PrestaShop/pull/14845

eternoendless replied:
Underscores are *NOT* valid and *WON'T* work.

eternoendless replied:
I mean, you can have underscores in your module name but you won't be able to translate it. Here's why:

Given a module `my_module`, its domain name would have to be either `MyModule` or `Mymodule`, because underscores are not allowed (at least not in that part). 

`MyModule` wouldn't be valid because internally domains are stored without dots, meaning that `MyModule` would be broken down into a subdomain like `My.Module`. `Mymodule` won't work either because we have no way of transforming `Mymodule` back into `my_module` to find the module directory and read its translation files.

mickaelandrieu commented on tests/Resources/modules/translationtest/views/templates/admin/index.html.twig:
```diff
@@ -4,7 +4,7 @@
     <div class="row">
         <div class="col-lg-8">
-            <h2>{{ 'Hello World'|trans({}, 'Modules.Translationtest.Translationtest') }}</h2>
+            <h2>{{ 'Hello World'|trans({}, 'Modules.Translationtest.Translation_test') }}</h2>
```
why do we have an underscore here? What is/are the rules to guess which domain name we should use in this case ?

LouiseBonnard replied:
Yep, same question.

eternoendless replied:
@LouiseBonnard You can ignore this, they are test resources.

@mickaelandrieu It belongs to the legacy fallback system, where you declare a domain and use already existing translations savec in the legacy format. In order for the "use existing translations" part to work, the domain name must follow some very specific rules, but basically: `Modules.<Nameofthemodule>.<Name_of_the_file_without_extension>`. This is so, specially the underscore part, because translations are stored that way in the legacy files.

Rest assured, this is all being documented and will be published soon.

mickaelandrieu commented on src/PrestaShopBundle/Translation/View/TreeBuilder.php:
```diff
@@ -137,11 +137,12 @@ public function makeTranslationsTree($catalogue)
         foreach ($catalogue as $domain => $messages) {
             $tableisedDomain = Inflector::tableize($domain);
-            $parts = array_reverse(explode('_', $tableisedDomain));
+            // the third component of the domain may have underscores, so we need to limit pieces to 3
```
can we export this business logic into a specific class/function to figure out what is done here ?

as a basic rule, everytime I want to comment a block of code, I ask myself: why not write a new function instead with a meaningful name and unit tests ?

This is not naive here, I've already contributed this code and this patch before (but the PR [was not merged](https://github.com/PrestaShop/PrestaShop/pull/13059/files#diff-f22a4655c95d20014e673956d4511214R141)).

eternoendless replied:
I think that's a different subject. This fix exists solely to allow underscores in the third part of a translation domain.

Other improvements can be done, for sure, although on a different PR.

mickaelandrieu reviewed (commented):
some minor comments

matthieu-rolland commented on classes/Context.php:
```diff
@@ -370,16 +371,22 @@ public function getTranslator()
 
-        $translator = $this->getTranslatorFromLocale($this->language->locale);
-        $this->translator = $translator;
+        $sfContainer = SymfonyContainer::getInstance();
```
You'll notice that I didn't use the **ContainerFinder**, the reason is that in its **getContainer** method, the symfony container comes last, so I ended up having another service container that doesn't have the translator service, even though the translator service is actually available via the Symfony container.

eternoendless replied:
That's weird, there shouldn't be more than one container in each environment. Maybe that's pointing to another bug somewhere else?

eternoendless replied:
Oh you meant in the front office? Then maybe we should declare it as a FO service, don't you think?

matthieu-rolland replied:
No I mean in back office, the ContainerFinder gives me an instance of **AdminContainer** instead of giving me a **SymfonyContainer**...

matthieu-rolland replied:
An issue has been created regarding this service container issue:

https://github.com/PrestaShop/PrestaShop/issues/14995

LouiseBonnard commented on tests/Resources/modules/translationtest/src/Controller/Admin/ExempleController.php:
```diff
@@ -24,15 +24,15 @@
     {
-        return $this->render('@Modules/translations/views/templates/admin/index.html.twig', [
+        return $this->render('@Modules/translationtest/views/templates/admin/index.html.twig', [
             'layoutTitle' => $this->trans('Modern controller', 'Modules.Translationtest.Admin'),
```
What is it?

PierreRambaud replied:
These, are tests resources, you can ignore everything in tests directory @LouiseBonnard  =)

LouiseBonnard commented on tests/Resources/modules/translationtest/views/templates/admin/someFile.with-things.tpl:
```diff
@@ -0,0 +1 @@
+<h1>{l s='Smarty template' mod='translationtest'}</h1>
```
Is it just for test purposes?

matthieu-rolland replied:
yes, this is the "translationtest" module, it's only for tests @LouiseBonnard

LouiseBonnard replied:
Okay, so it won't be part of the Crowdin catalog? I just want to be sure. :-)

matthieu-rolland replied:
Nope it shouldn't...  👍

eternoendless commented on src/PrestaShopBundle/Translation/TranslatorComponent.php:
```diff
@@ -29,6 +29,9 @@
 use Symfony\Component\Translation\Translator as BaseTranslatorComponent;
 
+/**
+ * Translator used by Context
```
If I'm not mistaken now it's only used in FO right?

```suggestion
 * Translator used in FO
```

matthieu-rolland committed:
revert the modifications affecting when the container is first called

matthieu-rolland commented:
I removed the part where I changed the place where the container was called for the first time, in
the **classes/controller/Controller.php**

The reason is that it creates a bug, only on PHP 5.6 when opcache is disabled.

![Capture d’écran 2019-08-06 à 15 15 20](https://user-images.githubusercontent.com/1784781/62542899-0eb93980-b85d-11e9-98c8-0834bed75184.png)

I know, it's weird, it looks a lot like this bug:

https://bugs.php.net/bug.php?id=66773

eternoendless commented:
> [2017-09-05 18:19 UTC] nikic@php.net
> This issue has been fixed in PHP 7 only.

It seems that they fixed it only in PHP 7 😅

eternoendless commented on classes/Context.php:
```diff
@@ -370,10 +371,16 @@ public function getTranslator()
-        return $translator;
+        if (null === $sfContainer) {
+            // symfony's container isn't available in front office, so we load and configure the translator component
+            $this->translator = $this->getTranslatorFromLocale($this->language->locale);
```
Can you confirm this method is _NEVER_ called in the BO context? (either legacy controller or symfony controller)

matthieu-rolland replied:
Yes, I carefully checked that we never enter this method in BO context, furthermore, I checked that the **TranslatorComponent** class is never instantiated in BO context.

The reason it works like this is that the SymfonyContainer is always available in BO context, in legacy and migrated pages alike.

matthieu-rolland replied:
Maybe it would be nice to improve the condition and check for BO context? I'm not sure about the proper way to do that though... ❓ 
I the mean time I like this condition, because internally, it is really the availability or not of the **SymfonyContainer** (and the symfony kernel) that implies which Translator will be used.

eternoendless replied:
Okay!

eternoendless reviewed (dismissed):
LGTM, QA label can be added once tests are green

khouloudbelguith commented:
Hi @matthieu-rolland,

I tried with different modules for example `productcomments` & `ps_mainmenu`.
The shop installed with an English language/France country & The employee language => French.
I would like to translate some expression's module from the back office in French => not translated.
In the FO => ok.
https://drive.google.com/file/d/16haO3ZAwdtaG0iP1PhzyRyiHG41s21Hr/view
When the employee's language is English => Translations => OK.
Thanks!

matthieu-rolland commented:
My mistake @khouloudbelguith , I should have waited before sending this PR to the QA...
Why? Because the BO module translation problem will be fixed in this PR from @eternoendless :
https://github.com/PrestaShop/PrestaShop/pull/14962

So I will wait until @eternoendless 's PR is merged, and then I will send this PR back to QA.

Sorry about this 🤦‍♂

matthieu-rolland committed:
fix installation and set locale for legacy back office page translation

matthieu-rolland commented on classes/Context.php:
```diff
@@ -364,16 +365,23 @@ public function updateCustomer(Customer $customer)
      * @return Translator
      */
-    public function getTranslator()
+    public function getTranslator($needComponentTranslator = false)
```
I made it possible to tell via a parameter which translator you need, because during installation it would get the framework bundle translator, which is configured to init the db sql loader, but the database is not ready yet, so it fails...

eternoendless replied:
I don't think this is a good idea, because

1. TranslatorComponent is meant to go away eventually
2. It exposes internal functionality
3. It's impossible tu understand if you don't know the underlying complexities 🗡 

If this is a special case just for the installer, let's just own it up:

```suggestion
    public function getTranslator($isInstaller = false)
```

or alternatively:

```suggestion
    public function getTranslator($withDatabase = true)
```

In any case, this parameter must be declared and explained in Phpdoc

matthieu-rolland replied:
done !

matthieu-rolland commented on classes/Context.php:
```diff
@@ -364,16 +365,23 @@ public function updateCustomer(Customer $customer)
+            $this->translator = $this->getTranslatorFromLocale($this->language->locale);
+        } else {
+            $this->translator = $sfContainer->get('translator');
+            $this->translator->setLocale($this->language->locale);
```
On legacy back office pages, the translator is called before listeners do their job of changing the locale.

For example the LegacyContext is called in the constructor of UserLocaleListener, and it instantiates AdminController who himself has translations done in its constructor. 

This is why we need to set the locale when initiating the translator, this is only necessary for legacy back office pages...

eternoendless replied:
It's better to add this comment in the code.

matthieu-rolland replied:
done !

matthieu-rolland committed:
minor improvements

- change getTranslator's parameter name
- add comments
- improve phpdoc

eternoendless reviewed (approved):
Looks good now, thanks!

matks commented:
Thank you @matthieu-rolland, @mickaelandrieu and @eternoendless
